kotlinx-coroutines-core/common/src/flow/SharedFlow.kt

```diff
@@ -532,36 +544,55 @@ internal open class SharedFlowImpl<T>(
         return index
     }
 
-    // Is called when a collector disappears or changes index, returns a list of continuations to resume after lock
+    /**
+     * Handle a collector having receiving the next value from the buffer or disappearing.
+     *
+     * As a result of both, the *index* of the collector--that is, the index of the next value in the buffer this
+     * collector intends to consume--changes, so the internal invariants need updating.
+     *
+     * In addition, it may turn out that either
+     * - This collector was the slowest one, and now that it has processed a value in the buffer,
+     *   the buffer has one more free slot to accommodate more values, so new emitters can be resumed.
+     * - This collector was the last one, and now that it's gone, emitters don't need to distribute values among
+     *   collectors anymore and can all continue executing immediately.
+     *
+     * In both cases, an array of continuations of the correct number of waiting emitters is returned
+     * so that they can be resumed.
+     */
     internal fun updateCollectorIndexLocked(oldIndex: Long): Array<Continuation<Unit>?> {
         assert { oldIndex >= minCollectorIndex }
         if (oldIndex > minCollectorIndex) return EMPTY_RESUMES // nothing changes, it was not min
-        // start computing new minimal index of active collectors
+        // start computing the new minimal index of active collectors
         val head = head
-        var newMinCollectorIndex = head + bufferSize
+        var newMinCollectorIndex = head + bufferSize // = this.bufferEndIndex
         // take into account a special case of sync shared flow that can go past 1st queued emitter
         if (bufferCapacity == 0 && queueSize > 0) newMinCollectorIndex++
         forEachSlotLocked { slot ->
-            @Suppress("ConvertTwoComparisonsToRangeCheck") // Bug in JS backend
-            if (slot.index >= 0 && slot.index < newMinCollectorIndex) newMinCollectorIndex = slot.index
+            if (slot.index in 0..<newMinCollectorIndex) newMinCollectorIndex = slot.index
         }
         assert { newMinCollectorIndex >= minCollectorIndex } // can only grow
         if (newMinCollectorIndex <= minCollectorIndex) return EMPTY_RESUMES // nothing changes
-        // Compute new buffer size if we drop items we no longer need and no emitter is resumed:
-        // We must keep all the items from newMinIndex to the end of buffer
-        var newBufferEndIndex = bufferEndIndex // var to grow when waiters are resumed
+        /* The new minimal resumer index has increased, so new emitters can be resumed,
```

Comment:
"resumer" and "minimal resumer index" are new terms, so it is not clear what they refer to. Given that it's only used once, avoid using it at all.

"Buffer space has freed up, so ..."